### PR TITLE
feature: filter visibility

### PR DESCRIPTION
**Pull request by Paul-Bob**
# Description

Fixes https://github.com/avo-hq/avo/issues/1253

# To do:

 - [x] Update the docs on actions and filters visible block
 - [x] Update the docs in the upgrade guide warning about the breaking change
 - [x] Add the `visible` block to the filter templates (generators). Maybe like on `dashboard.tt`?

# Breaking change

Check the visible block args!

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the [documentation](https://github.com/avo-hq/docs)
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by Paul-Bob** [comment]
@adrianthedev i think we should use a dry initializer on this. I inspired myself on how action visibility is done but i think we should refactor that one too, with the dry initializer both, action and filter visibility is more scalable.

**Comment by adrianthedev** [comment]
Ok @Paul-Bob. Let's do that then!

File: app/components/avo/actions_component.rb [hunk 1]
```diff
@@ -17,9 +17,7 @@ def render?
   end
 
   def actions
-    @actions.select do |action|
-      !action.class.in?(@exclude)
-    end
+    @actions.reject { |action| action.class.in?(@exclude) }
```
Comment (Paul-Bob) [inline app/components/avo/actions_component.rb:20]: unrelated

File: lib/avo/resources/controls/action.rb [hunk 2]
```diff
@@ -14,9 +14,7 @@ def initialize(klass, model: nil, resource: nil, view: nil, **args)
         end
 
         def action
-          return @instance if @instance.present?
-
-          @instance = @klass.new(model: @model, resource: @resource, view: @view)
+          @instance ||= @klass.new(model: @model, resource: @resource, view: @view)
```
Comment (Paul-Bob) [inline lib/avo/resources/controls/action.rb:17]: unrelated

File: spec/dummy/app/avo/actions/dummy_action.rb [hunk 3]
```diff
@@ -1,7 +1,18 @@
 class DummyAction < Avo::BaseAction
   self.name = "Dummy action"
   self.standalone = true
-  self.visible = ->(resource:, view:) do
+  self.visible = -> do
```
Comment (Paul-Bob) [inline spec/dummy/app/avo/actions/dummy_action.rb:4]: warning users about this change

File: lib/avo/base_action.rb [hunk 4]
```diff
@@ -56,8 +59,10 @@ def action_name
       self.class.to_s.demodulize.underscore.humanize(keep_id_suffix: true)
     end
 
-    def initialize(model: nil, resource: nil, user: nil, view: nil)
+    def initialize(model: nil, parent_model: nil, parent_resource: nil, resource: nil, user: nil, view: nil)
```
Comment (adrianthedev) [inline lib/avo/base_action.rb:62]: I'm a bit weary of adding that many things (`parent_model` and `parent_resource`) into an action. I feel they are a bit too extra.

Can we try not initializing with these two and (optionally) pass them in the `visible_in_view` method, so we have them in the Host?

File: lib/avo/resources/controls/action.rb (hunk 2 above)
Comment (adrianthedev) [reply lib/avo/resources/controls/action.rb:17]: 💪

File: spec/dummy/app/avo/actions/dummy_action.rb (hunk 3 above)
Comment (adrianthedev) [reply spec/dummy/app/avo/actions/dummy_action.rb:4]: Could you also update the docs and the upgrade guide?

**Review by adrianthedev — COMMENTED** [review]
Amazing job with the refactor!

I think we can lighten the load on `BaseAction` and `BaseFilter` and pass the parent stuff in the `visible_in_view` method.

I'll update the PR description with things still needed to be done.

File: spec/dummy/app/avo/actions/dummy_action.rb (hunk 3 above)
Comment (Paul-Bob) [reply spec/dummy/app/avo/actions/dummy_action.rb:4]: sure

File: lib/avo/base_action.rb (hunk 4 above)
Comment (Paul-Bob) [reply lib/avo/base_action.rb:62]: Great observation, do you think this apply to the filters too?

Comment (adrianthedev) [reply lib/avo/base_action.rb]: Yes, definitely!

Comment (Paul-Bob) [reply lib/avo/base_action.rb]: @adrianthedev refactor done and ready for last review and merge

**Review by adrianthedev — APPROVED** [review]
I left a comment about initializint the `BaseFilter` and these suggestions. Otherwise, the PR is good to go.

File: lib/avo/base_action.rb (hunk 4 above)
Comment (Paul-Bob) [reply lib/avo/base_action.rb]: sure, no problem
